File: AlgorithmStudy/8주차/BOJ2512/김재환2512.py
import sys

input = sys.stdin.readline

N = int(input().strip())
requests = list(map(int, input().split()))
M = int(input().strip())

# 예산 총합이 M 이하인 경우, 요청 최대값이 답
if sum(requests) <= M:
    print(max(requests))
else:
    low, high = 0, max(requests)
    answer = 0

    while low <= high:
        mid = (low + high) // 2  # 상한액 후보
        total = 0

        # 각 요청에 대해 cap 적용
        for r in requests:
            total += min(r, mid)

        # 총액이 M 이하이면 mid가 유효한 상한액
        if total <= M:
            answer = mid
            low = mid + 1
        else:
            high = mid - 1

    print(answer)

File: AlgorithmStudy/8주차/BOJ2512/김재환2512.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        int[] arr = new int[N];

        StringTokenizer st = new StringTokenizer(br.readLine());
        int max = 0;
        long sum = 0;

        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
            sum += arr[i];
            if (arr[i] > max) max = arr[i];
        }

        long M = Long.parseLong(br.readLine());

        // 모든 요청이 배정 가능하면 요청의 최대값
        if (sum <= M) {
            System.out.println(max);
            return;
        }

        int low = 0, high = max;
        int answer = 0;

        while (low <= high) {
            int mid = (low + high) / 2;
            long total = 0;

            for (int a : arr) {
                total += Math.min(a, mid);
            }

            if (total <= M) {
                answer = mid;
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        System.out.println(answer);
    }
}
